src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceWriter.cs

```diff
@@ -965,6 +974,21 @@ private void WriteRedType(TreeType node)
                     WriteLine("    {0} abstract {1}{2} {3} {{ get; }}", "public", (IsNew(field) ? "new " : ""), field.Type, field.Name);
                 }
 
+                var baseType = GetTreeType(node.Base);
+                if (baseType != null)
+                {
+                    var baseNodeFields = GetNodeOrNodeListFields(baseType);
+                    if (baseNodeFields.Count > 0)
+                    {
+                        WriteLine();
+                    }
+
+                    foreach (var baseField in baseNodeFields)
+                    {
+                        WriteLine($"    public new {node.Name} With{baseField.Name}({GetRedFieldType(baseField)} {CamelCase(baseField.Name)}) => ({node.Name})With{baseField.Name}Core({CamelCase(baseField.Name)});");
```

Comment:
Looking at `LambdaExpressionSyntax` and `AnonymousFunctionExpressionSyntax` as an example of this, I don't see why it's necessary. The base type already offer the `With` method, so what is the benefit of offering a new one here?

Reply:
Can you be more specific? Looking at AnonymousFunctionExpressoinSyntax, all i see is this:

```c#
  public abstract partial class AnonymousFunctionExpressionSyntax : ExpressionSyntax
  {
    internal AnonymousFunctionExpressionSyntax(Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.CSharpSyntaxNode green, SyntaxNode parent, int position)
      : base(green, parent, position)
    {
    }

    /// <summary>Gets the "async" token.</summary>
    public abstract SyntaxToken AsyncKeyword { get; }

    /// <summary>ExpressionSyntax or BlockSyntax representing the body of the lambda expression.</summary>
    public abstract CSharpSyntaxNode Body { get; }
  }
```

Reply:
I'm seeing the following.
Note comment at the bottom to point out my question.

```C#

  /// <summary>Provides the base class from which the classes that represent anonymous function expressions are derived.</summary>
  public abstract partial class AnonymousFunctionExpressionSyntax : ExpressionSyntax
  {
    internal AnonymousFunctionExpressionSyntax(Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.CSharpSyntaxNode green, SyntaxNode parent, int position)
      : base(green, parent, position)
    {
    }

    /// <summary>Gets the "async" token.</summary>
    public abstract SyntaxToken AsyncKeyword { get; }
    public AnonymousFunctionExpressionSyntax WithAsyncKeyword(SyntaxToken asyncKeyword) => WithAsyncKeywordCore(asyncKeyword);
    internal abstract AnonymousFunctionExpressionSyntax WithAsyncKeywordCore(SyntaxToken asyncKeyword);

    /// <summary>ExpressionSyntax or BlockSyntax representing the body of the lambda expression.</summary>
    public abstract CSharpSyntaxNode Body { get; }
    public AnonymousFunctionExpressionSyntax WithBody(CSharpSyntaxNode body) => WithBodyCore(body);
    internal abstract AnonymousFunctionExpressionSyntax WithBodyCore(CSharpSyntaxNode body);
  }

  /// <summary>Provides the base class from which the classes that represent lambda expressions are derived.</summary>
  public abstract partial class LambdaExpressionSyntax : AnonymousFunctionExpressionSyntax
  {
    internal LambdaExpressionSyntax(Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.CSharpSyntaxNode green, SyntaxNode parent, int position)
      : base(green, parent, position)
    {
    }

    /// <summary>SyntaxToken representing equals greater than.</summary>
    public abstract SyntaxToken ArrowToken { get; }
    public LambdaExpressionSyntax WithArrowToken(SyntaxToken arrowToken) => WithArrowTokenCore(arrowToken);
    internal abstract LambdaExpressionSyntax WithArrowTokenCore(SyntaxToken arrowToken);

// --------------------- these two methods are added, but seem un-necessary, since base already has such methods ----------------------------- //
    public new LambdaExpressionSyntax WithAsyncKeyword(SyntaxToken asyncKeyword) => (LambdaExpressionSyntax)WithAsyncKeywordCore(asyncKeyword);
    public new LambdaExpressionSyntax WithBody(CSharpSyntaxNode body) => (LambdaExpressionSyntax)WithBodyCore(body);
  }
```

---
In reply to: [183482677](https://github.com/dotnet/roslyn/pull/25945#discussion_r183482677) [](ancestors = 183482677)

Reply:
@jcouv Yes but if you know you have a LambdaExpressionSyntax, you want the more specific return type of LambdaExpressionSyntax but the based method only returns AnonymousFunctionExpressionSyntax.

Reply:
Ah! The return type is more specific. Got it. Thanks